---
layout: post
date: 2024-05-02
title: Iso-seq Note
tag: Daily tasks
categories: [Bioinformatics]
---

We have sequenced leaves from 25 plant species using PacBio Iso-Seq technology. Since this is my first time working with Iso-Seq data, I’ve documented key observations and insights gained during my exploration of these datasets.
<!--more-->

1. Pacbio does **NOT** equal to Iso-seq; PacBio is a sequencing technology platform developed by Pacific Biosciences, and it is not limited to Iso-Seq (Isoform Sequencing). PacBio is like Illumina; Illumina could do RNAseq and DNAseq, so does PacBio. When PacBio is used to do RNAseq, it is called Iso-seq.

2. **Barcode, primer and adapter**; According to my understanding, **barcodes** are used to tag different samples, and consequently are used for demultiplexed in downstream analysis. **Primers** refer to sequences used for PCR and converting RNA into cDNA. **Adapters** are short DNA sequences that are ligated or added to the ends of the DNA or RNA molecules during library preparation. In Iso-Seq library preparation, adapters are typically used to attach sequencing primers to the cDNA molecules.

3. **Full-pass and full-length**; **Full-pass reads**, also known as "single-pass", refer to reads generated from a single pass of the DNA polymerase around the DNA or RNA template. These reads are not subjected to multiple passes or sequencing rounds. My understanding is that a full-pass read should have adapters on both ends of the read; The following description is from --min-passes of lima docs.

   > 0 pass  : insert - adapter - insert \
   > 1 pass  : insert - adapter - INSERT - adapter - insert \
   > 2 passes: insert - adapter - INSERT - adapter - INSERT - adapter - insert

   We seem to talk about **full length** quite often. I think although it claims that the iso-seq is full length sequencing (from 3'end to 5'end), but this is not promised.

4. **CLR, subread, and CCS**; I think **CLR** (continuous length read) is **subread**. Subreads are the raw sequence data generated during PacBio sequencing. They represent the sequences from individual polymerase reads as they traverse the circular template multiple times. Subreads contain the raw data with any errors introduced during sequencing. They may vary in length and quality and are used as input data for various downstream applications, including error correction using CCS. **CCS** could be a algorithm that computes consensus sequences from multiple "passes" around a circularized single DNA molecule (SMRTbell template). The sequences after **CCS** is called Circular Consensus Sequencing (CCS) reads. They are also called HiFi reads (CCS Reads with predicted accuracy ≥ Q20), along with base quality values.

<figure style="display: flex; flex-direction: column; align-items: center; max-width: 100%; margin: 0 auto;">
  <img src="/assets/img/isoseq1.png">
</figure>

{:start="5"}

5. **ZMW**; A ZMW (Zero-Mode Waveguide) is a tiny well on the PacBio sequencing chip where individual DNA molecules are immobilized, and sequencing takes place. After CCS algorithm, there should only one ccs read per-ZMW. Although we could have forward strand and reverse strand for each molecule (ZMW), this would require higher sequencing depth. The one consensus attained for one ZMW is a represent of the molecule in the ZMW. I thought since reverse and forward strands should be reverse complementary to each other, knowing the forward strand would automatically tell us what the reverse strand looks like. It is not like that. According to my understanding, when sequencing the forward strand, some error happen, and this part has been missed. When sequencing the reverse strand, everything is fine. If we asked for forward and reverse strand consensuses separately, the forward strand will have this particular part missing; If we ask for one consensus only, this missing part would be made up from the reverse strand.

6. The “M” number designation indicates how many millions of ZMWs there are within that SMRT Cell. A SMRT Cell 8M contains 8 million ZMWs.

7. I noticed in the subreads BAM files, there are tags called "**ip**" and "**pw**". The "**ip**" tag stands for the "insertion/deletion quality profile." It provides information about the quality of the insertions and deletions within a subread. This tag is used to assess the quality of individual insertions and deletions within the sequence. The "ip" tag typically contains a list of quality values that correspond to the insertions and deletions at each position within the subread. These quality values help identify and evaluate the accuracy of insertion and deletion events. The "**pw**" stands for "pulse widths." It contains information about the duration of individual pulses recorded during the sequencing process for each base in a subread. Pulse widths provide information about the signal characteristics associated with each base, and they are used to improve the accuracy of base calling and to assess the quality of sequencing data. Variations in pulse width can be indicative of different chemical modifications, particularly DNA methylation, which is a common epigenetic modification.

8. The workflow I took is: 
<img src="/assets/img/isoseq-table.png" style="display: flex; flex-direction: column; align-items: center; max-width: 100%; margin: 0 auto;">

<img src="/assets/img/isoseq2.png" style="display: flex; flex-direction: column; align-items: center; max-width: 100%; margin: 0 auto;">

{:start="9"}

9. I thought the output after cluster is sequences of genes; given that 'Each consensus sequence in the output files from the clustering step typically represents a group of several full-length cDNAs that are highly similar to each other. These cDNAs are often clustered together because they likely correspond to different isoforms of the same gene'; But if one transcript (isoform) have been expressed several times, and captured in several ZMWs, the CCS from these ZMWs would clustered and represented by one consensus. And given that isoforms of one gene could be quite different from each other, the cluster step is likely to cluster copies of the same isoform into one consensus. Later, I found the --singletons of isoseq3, I am then pretty sure that the clusters should represent isoforms.

10. Iso-seq Primer-barcode: <https://www.pacb.com/wp-content/uploads/Shared-Protocol-PacBio-Barcodes-for-SMRT-Sequencing.pdf>; I think the construction of our library is like the following:
<img src="/assets/img/isoseq3.png" style="display: flex; flex-direction: column; align-items: center; max-width: 100%; margin: 0 auto;">


11. According to fastQC result (<mark>Overrepresented sequences</mark>), the construction of the read is 

   Barcode + Fwd primer + Insert -polyA + RC Rev primer + RC Barcode

   **Because Rev primer is on the 5' end of polyT in the <mark>Overrepresented sequences</mark>, so on the polyA strand, it should be RC Rev primer; on the polyT strand, the barcode is on the 5'end of Rev primer, so on the polyA strand, it should be RC barcode.**

   It does not matter the orientation of bc_3p in the barcode_primer.fa given to lima.

The way to construct the primer-barcode.fa for lima, is like the following:

>LotCor_5p \
>ACACACAGACTGTGAGGCAATGAAGTCGCAGGGTTGG 
>
>LotCor_3p     # <mark>It does not matter the orientation of bc_3p</mark> \
>ACACACAGACTGTGAGAAGCAGTGGTATCAACGCAGAGTAC

<img src="/assets/img/isoseq4.png" style="display: flex; flex-direction: column; align-items: center; max-width: 100%; margin: 0 auto;">

Also see <https://www.pacb.com/smrt-science/smrt-sequencing/multiplexing/>


<img src="/assets/img/isoseq5.png" style="display: flex; flex-direction: column; align-items: center; max-width: 100%; margin: 0 auto;">

{:start="12"}

12. Orientation of PacBio consensus transcripts are determined by the presence of polyA tails <https://www.pacb.com/wp-content/uploads/Isoform-Sequencing-Unveiling-the-Complexity-of-the-Eukaryotic-Transcriptome.pdf>

14. <img src="/assets/img/isoseq6.png" alt="image-20240717115658612" style="zoom:100%;" />